src/routes/documents/[id]-[slug]/sidebar/Actions.svelte

```diff
@@ -17,21 +20,40 @@
   import Flex from "$lib/components/common/Flex.svelte";
   import SidebarItem from "$lib/components/sidebar/SidebarItem.svelte";
 
+  // modals
+  import Edit from "$lib/components/forms/Edit.svelte";
+  import Reprocess from "$lib/components/forms/Reprocess.svelte";
+
   import { canonicalUrl, pdfUrl } from "$lib/api/documents";
 
   export let document: Document;
 
-  function relative(document: Document, path: string) {
-    return new URL(path, canonicalUrl(document)).href;
-  }
+  const modal: ModalContext = getContext("modal");
 
   // urls
-  $: edit = relative(document, "edit/");
   $: revisions = relative(document, "revisions/");
   $: pdf = pdfUrl(document).toString();
   $: annotate = relative(document, "annotate/");
   $: redact = relative(document, "redact/");
   $: modify = relative(document, "modify/");
+
+  function relative(document: Document, path: string) {
+    return new URL(path, canonicalUrl(document)).href;
+  }
+
+  function openEdit() {
+    $modal = {
+      component: Edit,
+      props: { document },
+    };
+  }
+
+  function openReprocess() {
+    $modal = {
+      component: Reprocess,
```

Comment:
The only issue I can see with this approach to modal rendering is that it breaks component composition. Am I correct about that? If so, could composition be maintained in a different approach?

Reply:
I think that's correct. For modals, we're sort of stepping out of the composition tree, or jumping back up it. We could create a helper function (or many) doing roughly the same thing, but I think that's messy. The only thing I don't really like about this approach is that it creates a cross-dependency between components that are otherwise unrelated, but I can't think of a way around that. This, at least, keeps the relationship explicit.

Reply:
React has a concept called [Portals](https://react.dev/reference/react-dom/createPortal) that enables rendering to a different part of the tree. While it doesn't look like Svelte has the same native functionality, there are community approaches to it:

- https://github.com/romkor/svelte-portal
- https://github.com/sveltejs/svelte/issues/3088#issuecomment-1065827485

Reply:
If I understand the way portals work in the two linked examples, opening a modal becomes a matter of maintaining some simple local state (is processing open? Render the modal. Otherwise, don't render it.) It would eliminate that cross-dependency you mentioned, while still allowing the modal to mount elsewhere in the tree.

Reply:
I'm gonna propose this work as a new issue—modal stuff is not too deeply rooted, so it can change outside this PR.